File: ios/FoodHomeApp/Telemetry/TelemetrySink.swift
import Foundation

protocol TelemetrySink {
    func record(_ event: TelemetryEvent)
}

struct NoOpTelemetrySink: TelemetrySink {
    func record(_ event: TelemetryEvent) {}
}

final class TelemetryReporter {
    private let sanitizer: TelemetrySanitizer
    private let sink: TelemetrySink

    init(sanitizer: TelemetrySanitizer, sink: TelemetrySink = NoOpTelemetrySink()) {
        self.sanitizer = sanitizer
        self.sink = sink
    }

    func record(
        _ name: TelemetryEventName,
        attributes: [String: String] = [:],
        routeURL: String? = nil
    ) {
        var values = attributes
        if let route = sanitizer.routeTemplate(routeURL) {
            values[TelemetryAttributeKey.routeTemplate.rawValue] = route
        }
        sink.record(sanitizer.sanitize(name: name, rawAttributes: values))
    }

    static func disabled(trustedOrigin: URL) -> TelemetryReporter {
        TelemetryReporter(
            sanitizer: TelemetrySanitizer(trustedOrigin: trustedOrigin),
            sink: NoOpTelemetrySink()
        )
    }
}

File: android/app/src/main/java/market/foodhome/app/telemetry/TelemetrySink.kt
package market.foodhome.app.telemetry

import java.net.URI

fun interface TelemetrySink {
    fun record(event: TelemetryEvent)
}

object NoOpTelemetrySink : TelemetrySink {
    override fun record(event: TelemetryEvent) = Unit
}

class TelemetryReporter(
    private val sanitizer: TelemetrySanitizer,
    private val sink: TelemetrySink = NoOpTelemetrySink,
) {
    fun record(
        name: TelemetryEventName,
        attributes: Map<String, String> = emptyMap(),
        routeUrl: String? = null,
    ) {
        val values = attributes.toMutableMap()
        sanitizer.routeTemplate(routeUrl)?.let {
            values[TelemetryAttributeKey.RouteTemplate.wireValue] = it
        }
        sink.record(sanitizer.sanitize(name, values))
    }

    companion object {
        fun disabled(trustedOrigin: URI): TelemetryReporter =
            TelemetryReporter(TelemetrySanitizer(trustedOrigin), NoOpTelemetrySink)
    }
}
